File: day03b.py
def check_slope(slope, horizontal_offset, vertical_offset):
    width = len(slope[0]) - 1
    height = len(slope)

    count = 0
    # For each row, use modular arithmetic to determine which column to consider
    for row in range(0, height, vertical_offset):
        col = (horizontal_offset * row // vertical_offset) % width
        if slope[row][col] == '#':
            count += 1
    return count

with open('day03.txt', 'r') as f:
    slope = f.readlines()
    result11 = check_slope(slope, 1, 1)
    result31 = check_slope(slope, 3, 1)
    result51 = check_slope(slope, 5, 1)
    result71 = check_slope(slope, 7, 1)
    result12 = check_slope(slope, 1, 2)
    print(result11 * result31 * result51 * result71 * result12)

File: day03/day03b.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

int check_slope(const std::vector<std::string> &slope, int horizontal_offset, int vertical_offset) {
  int width = slope[0].length();
  int height = slope.size();

  // For each row, use modular arithmetic to determine which column to consider
  int count = 0;
  for(int row = 0; row < height; row += vertical_offset) {
    int col = (horizontal_offset * row / vertical_offset) % width;
    if(slope[row][col] == '#')
      count++;
  }
  return count;
}

int main() {
  // Read input from file
  std::ifstream file("day03.txt");
  std::string line;
  std::vector<std::string> slope;

  while(std::getline(file, line)) {
    slope.push_back(line);
  }
  
  int result11 = check_slope(slope, 1, 1);
  int result31 = check_slope(slope, 3, 1);
  int result51 = check_slope(slope, 5, 1);
  int result71 = check_slope(slope, 7, 1);
  int result12 = check_slope(slope, 1, 2);
  std::cout << static_cast<long>(result11) * result31 * result51 * result71 * result12 << std::endl;

  return 0;
}
